{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'rvest':\n",
      "  method            from\n",
      "  read_xml.response xml2\n",
      "-- Attaching packages --------------------------------------- tidyverse 1.2.1 --\n",
      "v ggplot2 3.3.5     v purrr   0.3.4\n",
      "v tibble  3.1.6     v dplyr   1.0.7\n",
      "v tidyr   1.1.4     v stringr 1.4.0\n",
      "v readr   1.3.1     v forcats 0.4.0\n",
      "-- Conflicts ------------------------------------------ tidyverse_conflicts() --\n",
      "x dplyr::filter() masks stats::filter()\n",
      "x dplyr::lag()    masks stats::lag()\n",
      "\n",
      "Attaching package: 'lubridate'\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    date\n",
      "\n",
      "\n",
      "GOSemSim v2.12.0  For help: https://guangchuangyu.github.io/GOSemSim\n",
      "\n",
      "If you use GOSemSim in published research, please cite:\n",
      "Guangchuang Yu, Fei Li, Yide Qin, Xiaochen Bo, Yibo Wu, Shengqi Wang. GOSemSim: an R package for measuring semantic similarity among GO terms and gene products Bioinformatics 2010, 26(7):976-978\n",
      "Loading required package: AnnotationDbi\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from 'package:lubridate':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: 'Biobase'\n",
      "\n",
      "The following object is masked from 'package:httr':\n",
      "\n",
      "    content\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "The following objects are masked from 'package:lubridate':\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    first, rename\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    expand\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "The following object is masked from 'package:lubridate':\n",
      "\n",
      "    %within%\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "Attaching package: 'AnnotationDbi'\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    select\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(XML)\n",
    "library(httr)\n",
    "library(stringi)\n",
    "library(lubridate)\n",
    "library(GOSemSim)\n",
    "library(org.Hs.eg.db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# go term enrichment\n",
    "# inputs\n",
    "# required to include a list of genes for which to find go enrichments\n",
    "# optional to include a new gochart for comparing files, a new total_genes (for pval calc), and a path&name for saving outputs\n",
    "# notes: \n",
    "#      go table parsing can be found in create_go_table.ipynb\n",
    "\n",
    "gochart <- read.csv(\"gotable.csv\")\n",
    "go_enrichment_test <- function(set_of_genes, go_chart = gochart, total_genes = 24769, outputfile=NULL, v=TRUE){\n",
    "    go_testset <- go_chart[go_chart$genename %in% set_of_genes, ]\n",
    "    terms <- unique(go_testset$go_term)\n",
    "    if(v){\n",
    "        print(paste(\"TERMS: \", length(terms)), sep=\"\")\n",
    "    }\n",
    "    \n",
    "    enriched_terms <- c()\n",
    "    for (term in terms){ \n",
    "        genes_to_term_in_set <- nrow(go_testset[go_testset$go_term == term,])\n",
    "        genes_to_term_in_genome <- nrow(go_chart[go_chart$go_term == term,])\n",
    "        genes_not_term_in_genome <- total_genes - genes_to_term_in_genome\n",
    "        size_of_set <- length(set_of_genes)\n",
    "        test <- phyper(genes_to_term_in_set, genes_to_term_in_genome,\n",
    "                       genes_not_term_in_genome, size_of_set, lower.tail=FALSE)        \n",
    "        enriched_terms$term <- c(enriched_terms$term, term)\n",
    "        enriched_terms$pval <- c(enriched_terms$pval, test)\n",
    "        enriched_terms$genes_to_term_in_set <- c(enriched_terms$genes_to_term_in_set, genes_to_term_in_set)\n",
    "        enriched_terms$genes_to_term_in_genome <- c(enriched_terms$genes_to_term_in_genome, genes_to_term_in_genome)\n",
    "        enriched_terms$genes_not_term_in_genome <- c(enriched_terms$genes_not_term_in_genome, genes_not_term_in_genome)\n",
    "        enriched_terms$size_of_set <- c(enriched_terms$size_of_set, size_of_set)\n",
    "    }\n",
    "    \n",
    "    enriched_terms <- as.data.frame(enriched_terms)\n",
    "    enriched_terms <- enriched_terms %>% filter(genes_to_term_in_genome >= 10)\n",
    "    enriched_terms$p.adj <- p.adjust(enriched_terms$pval, method='fdr')\n",
    "    enriched_terms <- enriched_terms %>% filter(p.adj < 0.05)\n",
    "    \n",
    "    if(v){\n",
    "        print(paste(\"SIG TERMS:\", nrow(enriched_terms)))\n",
    "    }\n",
    "    \n",
    "    if(!is.null(outputfile)) {\n",
    "        write.csv(enriched_terms, outputfile)\n",
    "    }\n",
    "    return(enriched_terms)\n",
    "}   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"TERMS:  68\"\n",
      "[1] \"SIG TERMS: 24\"\n"
     ]
    }
   ],
   "source": [
    "# example of use\n",
    "filename_1 <- '/data/projects/julia.pratt/CS1_genomeSelection/all_rootstock_comparisons/Leaf_2018_Anthesis_Ungrafted-1103P.csv'\n",
    "data_1 <- read.csv(filename_1)\n",
    "genelist_1 <- data_1 %>% filter(padj < 0.05) %>% filter(log2FoldChange > 0)\n",
    "go_results_1 <- go_enrichment_test(genelist_1$X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"TERMS:  183\"\n",
      "[1] \"SIG TERMS: 23\"\n"
     ]
    }
   ],
   "source": [
    "# example of use\n",
    "filename_2 <- '/data/projects/julia.pratt/CS1_genomeSelection/all_rootstock_comparisons/Leaf_2018_Anthesis_Ungrafted-3309C.csv'\n",
    "data_2 <- read.csv(filename_2)\n",
    "genelist_2 <- data_2 %>% filter(padj < 0.05) %>% filter(log2FoldChange > 0)\n",
    "go_results_2 <- go_enrichment_test(genelist_2$X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# revigo\n",
    "\n",
    "# inputs: \n",
    "# required two matching lists: one of terms and the other of pvals\n",
    "# optional to specify a directory to store tmp files and a file path&name for saving output\n",
    "\n",
    "run_revigo <- function(terms, pvals, tmp_dir=\"./\", outputfile=NULL) {\n",
    "    tmp <- data.frame(\"term\"=terms, \"pval\"=pvals)\n",
    "    tmpfile <- paste(tmp_dir, \"tmp.txt\", sep=\"\")\n",
    "    write.table(tmp, tmpfile, quote=FALSE, row.names = FALSE, col.names = FALSE)\n",
    "    filedata <-readChar(tmpfile,file.info(tmpfile)$size)\n",
    "    \n",
    "    httr::POST(\n",
    "      url = \"http://revigo.irb.hr/Revigo.aspx\",\n",
    "      body = list(\n",
    "        cutoff = \"0.5\",\n",
    "        valueType = \"pvalue\",\n",
    "        speciesTaxon = \"0\",\n",
    "        measure = \"SIMREL\",\n",
    "        goList = filedata\n",
    "      ),\n",
    "      # application/x-www-form-urlencoded\n",
    "      encode = \"form\"\n",
    "    ) -> res\n",
    "\n",
    "    dat <- httr::content(res, encoding = \"UTF-8\")\n",
    "\n",
    "    # Write results to a file\n",
    "    dat <- stri_replace_all_fixed(dat, \"\\r\", \"\")\n",
    "    tmpfile <- paste(tmp_dir, \"tmpresults.html\", sep=\"\")\n",
    "    cat(dat, file=tmpfile, fill = FALSE)\n",
    "    resultsrevigo <- readHTMLTable(tmpfile)   \n",
    "    fullresultsrevigo <- setNames(data.frame(matrix(ncol = 4, nrow = 0)), c(\"Term.ID\", \"Name\", \"Value\", \"Category\"))\n",
    "    if(!is.null(resultsrevigo$BiologicalProcess)){\n",
    "        resultsrevigo$BiologicalProcess <- resultsrevigo$BiologicalProcess %>% \n",
    "                                            filter(Eliminated=='False') %>% \n",
    "                                            dplyr::select(`Term ID`, Name, Value)\n",
    "        resultsrevigo$BiologicalProcess['Category'] <- rep('BiologicalProcess', dim(resultsrevigo$BiologicalProcess)[1])\n",
    "        fullresultsrevigo <- rbind(fullresultsrevigo, resultsrevigo$BiologicalProcess)\n",
    "    }   \n",
    "    if(!is.null(resultsrevigo$CellularComponent)){\n",
    "        resultsrevigo$CellularComponent <- resultsrevigo$CellularComponent %>% \n",
    "                                            filter(Eliminated=='False') %>% \n",
    "                                            dplyr::select(`Term ID`, Name, Value)\n",
    "        resultsrevigo$CellularComponent['Category'] <- rep('CellularComponent', dim(resultsrevigo$CellularComponent)[1])\n",
    "        fullresultsrevigo <- rbind(fullresultsrevigo, resultsrevigo$CellularComponent)\n",
    "    }\n",
    "    if(!is.null(resultsrevigo$MolecularFunction)){\n",
    "        resultsrevigo$MolecularFunction <- resultsrevigo$MolecularFunction %>% \n",
    "                                            filter(Eliminated=='False') %>% \n",
    "                                            dplyr::select(`Term ID`, Name, Value)\n",
    "        resultsrevigo$MolecularFunction['Category'] <- rep('MolecularFunction', dim(resultsrevigo$MolecularFunction)[1])\n",
    "        fullresultsrevigo <- rbind(fullresultsrevigo, resultsrevigo$MolecularFunction)\n",
    "    }\n",
    "    \n",
    "    if(!is.null(outputfile)){\n",
    "        write.csv(data.frame(fullresultsrevigo), finalname, row.names=FALSE)\n",
    "    }\n",
    "    return(data.frame(fullresultsrevigo))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in stri_replace_all_fixed(dat, \"\\r\", \"\"):\n",
      "\"argument is not an atomic vector; coercing\""
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>Term.ID</th><th scope=col>Name</th><th scope=col>Value</th><th scope=col>Category</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>GO:0008272                                                                     </td><td>sulfate transport                                                              </td><td>-2.4316                                                                        </td><td>BiologicalProcess                                                              </td></tr>\n",
       "\t<tr><td>GO:0042545                                                                     </td><td>cell wall modification                                                         </td><td>-6.3890                                                                        </td><td>BiologicalProcess                                                              </td></tr>\n",
       "\t<tr><td>GO:0045944                                                                     </td><td>positive regulation of transcription by RNA polymerase II                      </td><td>-2.4306                                                                        </td><td>BiologicalProcess                                                              </td></tr>\n",
       "\t<tr><td>GO:0006073                                                                     </td><td>cellular glucan metabolic process                                              </td><td>-2.2031                                                                        </td><td>BiologicalProcess                                                              </td></tr>\n",
       "\t<tr><td>GO:0015629                                                                     </td><td>actin cytoskeleton                                                             </td><td>-5.4447                                                                        </td><td>CellularComponent                                                              </td></tr>\n",
       "\t<tr><td>GO:0005618                                                                     </td><td>cell wall                                                                      </td><td>-2.2031                                                                        </td><td>CellularComponent                                                              </td></tr>\n",
       "\t<tr><td>GO:0005576                                                                     </td><td>extracellular region                                                           </td><td>-3.9942                                                                        </td><td>CellularComponent                                                              </td></tr>\n",
       "\t<tr><td>GO:0004857                                                                     </td><td>enzyme inhibitor activity                                                      </td><td>-11.4355                                                                       </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0005509                                                                     </td><td>calcium ion binding                                                            </td><td>-4.1249                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0015116                                                                     </td><td>sulfate transmembrane transporter activity                                     </td><td>-2.4316                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0030599                                                                     </td><td>pectinesterase activity                                                        </td><td>-6.3890                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0016614                                                                     </td><td>oxidoreductase activity, acting on CH-OH group of donors                       </td><td>-2.4316                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0016757                                                                     </td><td>glycosyltransferase activity                                                   </td><td>-2.1834                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0000977                                                                     </td><td>RNA polymerase II transcription regulatory region sequence-specific DNA binding</td><td>-3.1844                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0003779                                                                     </td><td>actin binding                                                                  </td><td>-3.7481                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0004650                                                                     </td><td>polygalacturonase activity                                                     </td><td>-4.3374                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0016762                                                                     </td><td>xyloglucan:xyloglucosyl transferase activity                                   </td><td>-2.1214                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0008171                                                                     </td><td>O-methyltransferase activity                                                   </td><td>-2.1225                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0003700                                                                     </td><td>DNA-binding transcription factor activity                                      </td><td>-8.3998                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0003677                                                                     </td><td>DNA binding                                                                    </td><td>-2.4130                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "\t<tr><td>GO:0015267                                                                     </td><td>channel activity                                                               </td><td>-2.2460                                                                        </td><td>MolecularFunction                                                              </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llll}\n",
       " Term.ID & Name & Value & Category\\\\\n",
       "\\hline\n",
       "\t GO:0008272                                                                      & sulfate transport                                                               & -2.4316                                                                         & BiologicalProcess                                                              \\\\\n",
       "\t GO:0042545                                                                      & cell wall modification                                                          & -6.3890                                                                         & BiologicalProcess                                                              \\\\\n",
       "\t GO:0045944                                                                      & positive regulation of transcription by RNA polymerase II                       & -2.4306                                                                         & BiologicalProcess                                                              \\\\\n",
       "\t GO:0006073                                                                      & cellular glucan metabolic process                                               & -2.2031                                                                         & BiologicalProcess                                                              \\\\\n",
       "\t GO:0015629                                                                      & actin cytoskeleton                                                              & -5.4447                                                                         & CellularComponent                                                              \\\\\n",
       "\t GO:0005618                                                                      & cell wall                                                                       & -2.2031                                                                         & CellularComponent                                                              \\\\\n",
       "\t GO:0005576                                                                      & extracellular region                                                            & -3.9942                                                                         & CellularComponent                                                              \\\\\n",
       "\t GO:0004857                                                                      & enzyme inhibitor activity                                                       & -11.4355                                                                        & MolecularFunction                                                              \\\\\n",
       "\t GO:0005509                                                                      & calcium ion binding                                                             & -4.1249                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0015116                                                                      & sulfate transmembrane transporter activity                                      & -2.4316                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0030599                                                                      & pectinesterase activity                                                         & -6.3890                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0016614                                                                      & oxidoreductase activity, acting on CH-OH group of donors                        & -2.4316                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0016757                                                                      & glycosyltransferase activity                                                    & -2.1834                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0000977                                                                      & RNA polymerase II transcription regulatory region sequence-specific DNA binding & -3.1844                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0003779                                                                      & actin binding                                                                   & -3.7481                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0004650                                                                      & polygalacturonase activity                                                      & -4.3374                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0016762                                                                      & xyloglucan:xyloglucosyl transferase activity                                    & -2.1214                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0008171                                                                      & O-methyltransferase activity                                                    & -2.1225                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0003700                                                                      & DNA-binding transcription factor activity                                       & -8.3998                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0003677                                                                      & DNA binding                                                                     & -2.4130                                                                         & MolecularFunction                                                              \\\\\n",
       "\t GO:0015267                                                                      & channel activity                                                                & -2.2460                                                                         & MolecularFunction                                                              \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| Term.ID | Name | Value | Category |\n",
       "|---|---|---|---|\n",
       "| GO:0008272                                                                      | sulfate transport                                                               | -2.4316                                                                         | BiologicalProcess                                                               |\n",
       "| GO:0042545                                                                      | cell wall modification                                                          | -6.3890                                                                         | BiologicalProcess                                                               |\n",
       "| GO:0045944                                                                      | positive regulation of transcription by RNA polymerase II                       | -2.4306                                                                         | BiologicalProcess                                                               |\n",
       "| GO:0006073                                                                      | cellular glucan metabolic process                                               | -2.2031                                                                         | BiologicalProcess                                                               |\n",
       "| GO:0015629                                                                      | actin cytoskeleton                                                              | -5.4447                                                                         | CellularComponent                                                               |\n",
       "| GO:0005618                                                                      | cell wall                                                                       | -2.2031                                                                         | CellularComponent                                                               |\n",
       "| GO:0005576                                                                      | extracellular region                                                            | -3.9942                                                                         | CellularComponent                                                               |\n",
       "| GO:0004857                                                                      | enzyme inhibitor activity                                                       | -11.4355                                                                        | MolecularFunction                                                               |\n",
       "| GO:0005509                                                                      | calcium ion binding                                                             | -4.1249                                                                         | MolecularFunction                                                               |\n",
       "| GO:0015116                                                                      | sulfate transmembrane transporter activity                                      | -2.4316                                                                         | MolecularFunction                                                               |\n",
       "| GO:0030599                                                                      | pectinesterase activity                                                         | -6.3890                                                                         | MolecularFunction                                                               |\n",
       "| GO:0016614                                                                      | oxidoreductase activity, acting on CH-OH group of donors                        | -2.4316                                                                         | MolecularFunction                                                               |\n",
       "| GO:0016757                                                                      | glycosyltransferase activity                                                    | -2.1834                                                                         | MolecularFunction                                                               |\n",
       "| GO:0000977                                                                      | RNA polymerase II transcription regulatory region sequence-specific DNA binding | -3.1844                                                                         | MolecularFunction                                                               |\n",
       "| GO:0003779                                                                      | actin binding                                                                   | -3.7481                                                                         | MolecularFunction                                                               |\n",
       "| GO:0004650                                                                      | polygalacturonase activity                                                      | -4.3374                                                                         | MolecularFunction                                                               |\n",
       "| GO:0016762                                                                      | xyloglucan:xyloglucosyl transferase activity                                    | -2.1214                                                                         | MolecularFunction                                                               |\n",
       "| GO:0008171                                                                      | O-methyltransferase activity                                                    | -2.1225                                                                         | MolecularFunction                                                               |\n",
       "| GO:0003700                                                                      | DNA-binding transcription factor activity                                       | -8.3998                                                                         | MolecularFunction                                                               |\n",
       "| GO:0003677                                                                      | DNA binding                                                                     | -2.4130                                                                         | MolecularFunction                                                               |\n",
       "| GO:0015267                                                                      | channel activity                                                                | -2.2460                                                                         | MolecularFunction                                                               |\n",
       "\n"
      ],
      "text/plain": [
       "   Term.ID   \n",
       "1  GO:0008272\n",
       "2  GO:0042545\n",
       "3  GO:0045944\n",
       "4  GO:0006073\n",
       "5  GO:0015629\n",
       "6  GO:0005618\n",
       "7  GO:0005576\n",
       "8  GO:0004857\n",
       "9  GO:0005509\n",
       "10 GO:0015116\n",
       "11 GO:0030599\n",
       "12 GO:0016614\n",
       "13 GO:0016757\n",
       "14 GO:0000977\n",
       "15 GO:0003779\n",
       "16 GO:0004650\n",
       "17 GO:0016762\n",
       "18 GO:0008171\n",
       "19 GO:0003700\n",
       "20 GO:0003677\n",
       "21 GO:0015267\n",
       "   Name                                                                           \n",
       "1  sulfate transport                                                              \n",
       "2  cell wall modification                                                         \n",
       "3  positive regulation of transcription by RNA polymerase II                      \n",
       "4  cellular glucan metabolic process                                              \n",
       "5  actin cytoskeleton                                                             \n",
       "6  cell wall                                                                      \n",
       "7  extracellular region                                                           \n",
       "8  enzyme inhibitor activity                                                      \n",
       "9  calcium ion binding                                                            \n",
       "10 sulfate transmembrane transporter activity                                     \n",
       "11 pectinesterase activity                                                        \n",
       "12 oxidoreductase activity, acting on CH-OH group of donors                       \n",
       "13 glycosyltransferase activity                                                   \n",
       "14 RNA polymerase II transcription regulatory region sequence-specific DNA binding\n",
       "15 actin binding                                                                  \n",
       "16 polygalacturonase activity                                                     \n",
       "17 xyloglucan:xyloglucosyl transferase activity                                   \n",
       "18 O-methyltransferase activity                                                   \n",
       "19 DNA-binding transcription factor activity                                      \n",
       "20 DNA binding                                                                    \n",
       "21 channel activity                                                               \n",
       "   Value    Category         \n",
       "1  -2.4316  BiologicalProcess\n",
       "2  -6.3890  BiologicalProcess\n",
       "3  -2.4306  BiologicalProcess\n",
       "4  -2.2031  BiologicalProcess\n",
       "5  -5.4447  CellularComponent\n",
       "6  -2.2031  CellularComponent\n",
       "7  -3.9942  CellularComponent\n",
       "8  -11.4355 MolecularFunction\n",
       "9  -4.1249  MolecularFunction\n",
       "10 -2.4316  MolecularFunction\n",
       "11 -6.3890  MolecularFunction\n",
       "12 -2.4316  MolecularFunction\n",
       "13 -2.1834  MolecularFunction\n",
       "14 -3.1844  MolecularFunction\n",
       "15 -3.7481  MolecularFunction\n",
       "16 -4.3374  MolecularFunction\n",
       "17 -2.1214  MolecularFunction\n",
       "18 -2.1225  MolecularFunction\n",
       "19 -8.3998  MolecularFunction\n",
       "20 -2.4130  MolecularFunction\n",
       "21 -2.2460  MolecularFunction"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# example of use, see also go_enrichment_test\n",
    "run_revigo(terms=go_results_2[[\"term\"]], pvals=go_results_2[[\"pval\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ipr enrichment\n",
    "# inputs:\n",
    "\n",
    "\n",
    "# scanned_file <- read.csv('/data/projects/julia.pratt/CS1_genomeSelection/scripts/vitvi.vcostv3.clean.pep.txt', sep='\\t')\n",
    "ipr_enrichment_test <- function(set_of_genes, scan=scanned_file, v=TRUE){\n",
    "    ipr_sub <- scan %>%\n",
    "    dplyr::select(genename, e.value, ipr_accession) %>%\n",
    "    dplyr::filter(e.value < 1e-05) %>% \n",
    "    dplyr::filter(ipr_accession != '') %>%\n",
    "    dplyr::select(genename, ipr_accession) %>%\n",
    "    unique()\n",
    "    \n",
    "    set_of_genes <- set_of_genes %>% paste('.t01', sep='')\n",
    "    ipr_testset <- ipr_sub[ipr_sub$genename %in% set_of_genes,]\n",
    "    terms <- unique(ipr_testset$ipr_accession)\n",
    "    \n",
    "    if(v){\n",
    "        print(str_interp(\"analyzing ${length(terms)} IPR accessions...\"))\n",
    "    }\n",
    "    \n",
    "    enriched_terms <- c() \n",
    "    for (term in terms){\n",
    "        genes_to_term_in_set <- nrow(ipr_testset[ipr_testset$ipr_accession == term,])\n",
    "        genes_to_term_in_genome <- nrow(ipr_sub[ipr_sub$ipr_accession == term,])\n",
    "        genes_not_term_in_genome <- 24769 - genes_to_term_in_genome\n",
    "        size_of_set <- length(set_of_genes)\n",
    "        test <- phyper(genes_to_term_in_set, genes_to_term_in_genome,\n",
    "                       genes_not_term_in_genome, size_of_set, lower.tail=FALSE)\n",
    "        \n",
    "        if (test < 0.05/length(terms) & \n",
    "            genes_to_term_in_genome >= 10 & \n",
    "            genes_to_term_in_set >= 3){\n",
    "            if (v) {\n",
    "                print(c(term, test, genes_to_term_in_set, genes_to_term_in_genome))\n",
    "            }\n",
    "            enriched_terms$term <- c(enriched_terms$term, term)\n",
    "            enriched_terms$genes_to_term_in_set <- c(enriched_terms$genes_to_term_in_set, genes_to_term_in_set)\n",
    "            enriched_terms$genes_to_term_in_genome <- c(enriched_terms$genes_to_term_in_genome, genes_to_term_in_genome)\n",
    "            enriched_terms$pval <- c(enriched_terms$pval, test)\n",
    "        }\n",
    "    }\n",
    "    return(enriched_terms)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"analyzing 154 IPR accessions...\"\n",
      "[1] \"IPR001471\"            \"1.07609723465595e-17\" \"15\"                  \n",
      "[4] \"147\"                 \n",
      "[1] \"IPR036955\"            \"1.20193093598421e-17\" \"15\"                  \n",
      "[4] \"148\"                 \n",
      "[1] \"IPR016177\"            \"3.85411345356568e-17\" \"15\"                  \n",
      "[4] \"159\"                 \n",
      "[1] \"IPR002048\"           \"4.2353739574481e-07\" \"7\"                  \n",
      "[4] \"145\"                \n",
      "[1] \"IPR011992\"            \"4.39768215039543e-06\" \"7\"                   \n",
      "[4] \"198\"                 \n",
      "[1] \"IPR000757\"            \"4.89451036565296e-05\" \"3\"                   \n",
      "[4] \"43\"                  \n",
      "[1] \"IPR010713\"            \"2.98084961180961e-05\" \"3\"                   \n",
      "[4] \"38\"                  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$term</dt>\n",
       "\t\t<dd><ol class=list-inline>\n",
       "\t<li>'IPR001471'</li>\n",
       "\t<li>'IPR036955'</li>\n",
       "\t<li>'IPR016177'</li>\n",
       "\t<li>'IPR002048'</li>\n",
       "\t<li>'IPR011992'</li>\n",
       "\t<li>'IPR000757'</li>\n",
       "\t<li>'IPR010713'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$genes_to_term_in_set</dt>\n",
       "\t\t<dd><ol class=list-inline>\n",
       "\t<li>15</li>\n",
       "\t<li>15</li>\n",
       "\t<li>15</li>\n",
       "\t<li>7</li>\n",
       "\t<li>7</li>\n",
       "\t<li>3</li>\n",
       "\t<li>3</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$genes_to_term_in_genome</dt>\n",
       "\t\t<dd><ol class=list-inline>\n",
       "\t<li>147</li>\n",
       "\t<li>148</li>\n",
       "\t<li>159</li>\n",
       "\t<li>145</li>\n",
       "\t<li>198</li>\n",
       "\t<li>43</li>\n",
       "\t<li>38</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$pval</dt>\n",
       "\t\t<dd><ol class=list-inline>\n",
       "\t<li>1.07609723465595e-17</li>\n",
       "\t<li>1.20193093598421e-17</li>\n",
       "\t<li>3.85411345356568e-17</li>\n",
       "\t<li>4.2353739574481e-07</li>\n",
       "\t<li>4.39768215039543e-06</li>\n",
       "\t<li>4.89451036565296e-05</li>\n",
       "\t<li>2.98084961180961e-05</li>\n",
       "</ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$term] \\begin{enumerate*}\n",
       "\\item 'IPR001471'\n",
       "\\item 'IPR036955'\n",
       "\\item 'IPR016177'\n",
       "\\item 'IPR002048'\n",
       "\\item 'IPR011992'\n",
       "\\item 'IPR000757'\n",
       "\\item 'IPR010713'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$genes\\_to\\_term\\_in\\_set] \\begin{enumerate*}\n",
       "\\item 15\n",
       "\\item 15\n",
       "\\item 15\n",
       "\\item 7\n",
       "\\item 7\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$genes\\_to\\_term\\_in\\_genome] \\begin{enumerate*}\n",
       "\\item 147\n",
       "\\item 148\n",
       "\\item 159\n",
       "\\item 145\n",
       "\\item 198\n",
       "\\item 43\n",
       "\\item 38\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$pval] \\begin{enumerate*}\n",
       "\\item 1.07609723465595e-17\n",
       "\\item 1.20193093598421e-17\n",
       "\\item 3.85411345356568e-17\n",
       "\\item 4.2353739574481e-07\n",
       "\\item 4.39768215039543e-06\n",
       "\\item 4.89451036565296e-05\n",
       "\\item 2.98084961180961e-05\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$term\n",
       ":   1. 'IPR001471'\n",
       "2. 'IPR036955'\n",
       "3. 'IPR016177'\n",
       "4. 'IPR002048'\n",
       "5. 'IPR011992'\n",
       "6. 'IPR000757'\n",
       "7. 'IPR010713'\n",
       "\n",
       "\n",
       "\n",
       "$genes_to_term_in_set\n",
       ":   1. 15\n",
       "2. 15\n",
       "3. 15\n",
       "4. 7\n",
       "5. 7\n",
       "6. 3\n",
       "7. 3\n",
       "\n",
       "\n",
       "\n",
       "$genes_to_term_in_genome\n",
       ":   1. 147\n",
       "2. 148\n",
       "3. 159\n",
       "4. 145\n",
       "5. 198\n",
       "6. 43\n",
       "7. 38\n",
       "\n",
       "\n",
       "\n",
       "$pval\n",
       ":   1. 1.07609723465595e-17\n",
       "2. 1.20193093598421e-17\n",
       "3. 3.85411345356568e-17\n",
       "4. 4.2353739574481e-07\n",
       "5. 4.39768215039543e-06\n",
       "6. 4.89451036565296e-05\n",
       "7. 2.98084961180961e-05\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$term\n",
       "[1] \"IPR001471\" \"IPR036955\" \"IPR016177\" \"IPR002048\" \"IPR011992\" \"IPR000757\"\n",
       "[7] \"IPR010713\"\n",
       "\n",
       "$genes_to_term_in_set\n",
       "[1] 15 15 15  7  7  3  3\n",
       "\n",
       "$genes_to_term_in_genome\n",
       "[1] 147 148 159 145 198  43  38\n",
       "\n",
       "$pval\n",
       "[1] 1.076097e-17 1.201931e-17 3.854113e-17 4.235374e-07 4.397682e-06\n",
       "[6] 4.894510e-05 2.980850e-05\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# example of use\n",
    "ipr_results <- ipr_enrichment_test(as.character(genelist_1$X))\n",
    "ipr_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pfam enrichment\n",
    "# inputs:\n",
    "# outputs: \n",
    "\n",
    "scan <- read.csv('/data/projects/julia.pratt/CS1_genomeSelection/scripts/vitvi.vcostv3.clean.pep.txt', sep='\\t')\n",
    "\n",
    "# scanned_file <- read.csv('/data/projects/julia.pratt/CS1_genomeSelection/scripts/vitvi.vcostv3.clean.pep.txt', sep='\\t')\n",
    "pfam_enrichment_test <- function(set_of_genes, scan=scanned_file, v=TRUE){\n",
    "    pfam_sub <- scan %>%\n",
    "    dplyr::select(genename, analysis, signature_accession, signature_description, e.value, ipr_accession) %>%\n",
    "    dplyr::filter(e.value < 1e-05) %>% \n",
    "    dplyr::filter(analysis == 'Pfam') %>%\n",
    "    dplyr::select(genename, signature_accession, signature_description, ipr_accession) %>%\n",
    "    unique()\n",
    "    \n",
    "    set_of_genes <- set_of_genes %>% paste('.t01', sep='')\n",
    "    pfam_testset <- pfam_sub[pfam_sub$genename %in% set_of_genes,]\n",
    "    terms <- unique(pfam_testset$signature_accession)\n",
    "    \n",
    "    if(v){\n",
    "        print(str_interp(\"analyzing ${length(terms)} Pfam domains...\"))\n",
    "    }\n",
    "    \n",
    "    enriched_terms <- c() \n",
    "    for (term in terms){\n",
    "        genes_to_term_in_set <- nrow(pfam_testset[pfam_testset$signature_accession == term,])\n",
    "        genes_to_term_in_genome <- nrow(pfam_sub[pfam_sub$signature_accession == term,])\n",
    "        genes_not_term_in_genome <- 24769 - genes_to_term_in_genome\n",
    "        size_of_set <- length(set_of_genes)\n",
    "        test <- phyper(genes_to_term_in_set, genes_to_term_in_genome,\n",
    "                       genes_not_term_in_genome, size_of_set, lower.tail=FALSE)\n",
    "        \n",
    "        if (test < 0.05/length(terms) & \n",
    "            genes_to_term_in_genome >= 10 & \n",
    "            genes_to_term_in_set >= 3){\n",
    "            if (v) {\n",
    "                print(c(term, test, genes_to_term_in_set, genes_to_term_in_genome))\n",
    "            }\n",
    "            enriched_terms$term <- c(enriched_terms$term, term)\n",
    "            enriched_terms$genes_to_term_in_set <- c(enriched_terms$genes_to_term_in_set, genes_to_term_in_set)\n",
    "            enriched_terms$genes_to_term_in_genome <- c(enriched_terms$genes_to_term_in_genome, genes_to_term_in_genome)\n",
    "            enriched_terms$pval <- c(enriched_terms$pval, test)\n",
    "        }\n",
    "    }\n",
    "    return(enriched_terms)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"analyzing 85 Pfam domains...\"\n",
      "[1] \"PF00847\"              \"6.11267309335109e-18\" \"15\"                  \n",
      "[4] \"142\"                 \n",
      "[1] \"PF13499\"             \"9.3548808237208e-07\" \"6\"                  \n",
      "[4] \"111\"                \n",
      "[1] \"PF00722\"              \"4.89451036565296e-05\" \"3\"                   \n",
      "[4] \"43\"                  \n",
      "[1] \"PF06955\"              \"2.98084961180961e-05\" \"3\"                   \n",
      "[4] \"38\"                  \n",
      "[1] \"PF13833\"              \"5.01617790663546e-09\" \"5\"                   \n",
      "[4] \"30\"                  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$term</dt>\n",
       "\t\t<dd><ol class=list-inline>\n",
       "\t<li>'PF00847'</li>\n",
       "\t<li>'PF13499'</li>\n",
       "\t<li>'PF00722'</li>\n",
       "\t<li>'PF06955'</li>\n",
       "\t<li>'PF13833'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$genes_to_term_in_set</dt>\n",
       "\t\t<dd><ol class=list-inline>\n",
       "\t<li>15</li>\n",
       "\t<li>6</li>\n",
       "\t<li>3</li>\n",
       "\t<li>3</li>\n",
       "\t<li>5</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$genes_to_term_in_genome</dt>\n",
       "\t\t<dd><ol class=list-inline>\n",
       "\t<li>142</li>\n",
       "\t<li>111</li>\n",
       "\t<li>43</li>\n",
       "\t<li>38</li>\n",
       "\t<li>30</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$pval</dt>\n",
       "\t\t<dd><ol class=list-inline>\n",
       "\t<li>6.11267309335109e-18</li>\n",
       "\t<li>9.3548808237208e-07</li>\n",
       "\t<li>4.89451036565296e-05</li>\n",
       "\t<li>2.98084961180961e-05</li>\n",
       "\t<li>5.01617790663546e-09</li>\n",
       "</ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$term] \\begin{enumerate*}\n",
       "\\item 'PF00847'\n",
       "\\item 'PF13499'\n",
       "\\item 'PF00722'\n",
       "\\item 'PF06955'\n",
       "\\item 'PF13833'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$genes\\_to\\_term\\_in\\_set] \\begin{enumerate*}\n",
       "\\item 15\n",
       "\\item 6\n",
       "\\item 3\n",
       "\\item 3\n",
       "\\item 5\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$genes\\_to\\_term\\_in\\_genome] \\begin{enumerate*}\n",
       "\\item 142\n",
       "\\item 111\n",
       "\\item 43\n",
       "\\item 38\n",
       "\\item 30\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$pval] \\begin{enumerate*}\n",
       "\\item 6.11267309335109e-18\n",
       "\\item 9.3548808237208e-07\n",
       "\\item 4.89451036565296e-05\n",
       "\\item 2.98084961180961e-05\n",
       "\\item 5.01617790663546e-09\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$term\n",
       ":   1. 'PF00847'\n",
       "2. 'PF13499'\n",
       "3. 'PF00722'\n",
       "4. 'PF06955'\n",
       "5. 'PF13833'\n",
       "\n",
       "\n",
       "\n",
       "$genes_to_term_in_set\n",
       ":   1. 15\n",
       "2. 6\n",
       "3. 3\n",
       "4. 3\n",
       "5. 5\n",
       "\n",
       "\n",
       "\n",
       "$genes_to_term_in_genome\n",
       ":   1. 142\n",
       "2. 111\n",
       "3. 43\n",
       "4. 38\n",
       "5. 30\n",
       "\n",
       "\n",
       "\n",
       "$pval\n",
       ":   1. 6.11267309335109e-18\n",
       "2. 9.3548808237208e-07\n",
       "3. 4.89451036565296e-05\n",
       "4. 2.98084961180961e-05\n",
       "5. 5.01617790663546e-09\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$term\n",
       "[1] \"PF00847\" \"PF13499\" \"PF00722\" \"PF06955\" \"PF13833\"\n",
       "\n",
       "$genes_to_term_in_set\n",
       "[1] 15  6  3  3  5\n",
       "\n",
       "$genes_to_term_in_genome\n",
       "[1] 142 111  43  38  30\n",
       "\n",
       "$pval\n",
       "[1] 6.112673e-18 9.354881e-07 4.894510e-05 2.980850e-05 5.016178e-09\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# example of use\n",
    "pfam_results <- pfam_enrichment_test(as.character(genelist_1$X))\n",
    "pfam_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gosemsim\n",
    "# inputs: two lists of go terms, as characters, filtered by ontology (passed in as \"BP\", \"CC\", \"MF\")\n",
    "# measurement recommended (default): \"Wang\"\n",
    "\n",
    "compare_two <- function(go1, go2, ont, measure=\"Wang\") {\n",
    "    hsGO <- godata('org.Hs.eg.db', ont=ont)\n",
    "    measure <- mgoSim(go1, go2, semData=hsGO, measure=measure)\n",
    "    gosemsims <- rbind(gosemsims, data.frame(\n",
    "        \"filename\" = f,\n",
    "        \"ont\" = ont,\n",
    "        \"semsim\" = measure)\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>term</th><th scope=col>pval</th><th scope=col>genes_to_term_in_set</th><th scope=col>genes_to_term_in_genome</th><th scope=col>genes_not_term_in_genome</th><th scope=col>size_of_set</th><th scope=col>p.adj</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>GO:0004568  </td><td>5.158789e-03</td><td> 1          </td><td>  23        </td><td>24746       </td><td>116         </td><td>1.418667e-02</td></tr>\n",
       "\t<tr><td>GO:0006032  </td><td>5.158789e-03</td><td> 1          </td><td>  23        </td><td>24746       </td><td>116         </td><td>1.418667e-02</td></tr>\n",
       "\t<tr><td>GO:0016998  </td><td>5.158789e-03</td><td> 1          </td><td>  23        </td><td>24746       </td><td>116         </td><td>1.418667e-02</td></tr>\n",
       "\t<tr><td>GO:0006334  </td><td>1.907271e-03</td><td> 1          </td><td>  14        </td><td>24755       </td><td>116         </td><td>9.536357e-03</td></tr>\n",
       "\t<tr><td>GO:0003677  </td><td>2.781172e-06</td><td>17          </td><td>1080        </td><td>23689       </td><td>116         </td><td>3.824112e-05</td></tr>\n",
       "\t<tr><td>GO:0003700  </td><td>8.756623e-13</td><td>16          </td><td> 355        </td><td>24414       </td><td>116         </td><td>4.816142e-11</td></tr>\n",
       "\t<tr><td>GO:0006355  </td><td>1.055418e-09</td><td>17          </td><td> 638        </td><td>24131       </td><td>116         </td><td>2.902399e-08</td></tr>\n",
       "\t<tr><td>GO:0006457  </td><td>3.855627e-03</td><td> 2          </td><td>  67        </td><td>24702       </td><td>116         </td><td>1.413730e-02</td></tr>\n",
       "\t<tr><td>GO:0051082  </td><td>2.199133e-03</td><td> 2          </td><td>  55        </td><td>24714       </td><td>116         </td><td>1.007936e-02</td></tr>\n",
       "\t<tr><td>GO:0005509  </td><td>2.167969e-06</td><td> 7          </td><td> 180        </td><td>24589       </td><td>116         </td><td>3.824112e-05</td></tr>\n",
       "\t<tr><td>GO:0005388  </td><td>7.070310e-03</td><td> 1          </td><td>  27        </td><td>24742       </td><td>116         </td><td>1.690726e-02</td></tr>\n",
       "\t<tr><td>GO:0070588  </td><td>6.566561e-03</td><td> 1          </td><td>  26        </td><td>24743       </td><td>116         </td><td>1.641640e-02</td></tr>\n",
       "\t<tr><td>GO:0005618  </td><td>2.676722e-05</td><td> 3          </td><td>  37        </td><td>24732       </td><td>116         </td><td>2.453662e-04</td></tr>\n",
       "\t<tr><td>GO:0006073  </td><td>2.676722e-05</td><td> 3          </td><td>  37        </td><td>24732       </td><td>116         </td><td>2.453662e-04</td></tr>\n",
       "\t<tr><td>GO:0016762  </td><td>3.309550e-05</td><td> 3          </td><td>  39        </td><td>24730       </td><td>116         </td><td>2.600361e-04</td></tr>\n",
       "\t<tr><td>GO:0048046  </td><td>1.571080e-03</td><td> 3          </td><td> 106        </td><td>24663       </td><td>116         </td><td>8.640942e-03</td></tr>\n",
       "\t<tr><td>GO:0010411  </td><td>5.387248e-04</td><td> 2          </td><td>  34        </td><td>24735       </td><td>116         </td><td>3.703733e-03</td></tr>\n",
       "\t<tr><td>GO:0042546  </td><td>6.384324e-04</td><td> 2          </td><td>  36        </td><td>24733       </td><td>116         </td><td>3.901532e-03</td></tr>\n",
       "\t<tr><td>GO:0016779  </td><td>4.724593e-03</td><td> 1          </td><td>  22        </td><td>24747       </td><td>116         </td><td>1.418667e-02</td></tr>\n",
       "\t<tr><td>GO:0008237  </td><td>5.610630e-03</td><td> 1          </td><td>  24        </td><td>24745       </td><td>116         </td><td>1.469451e-02</td></tr>\n",
       "\t<tr><td>GO:0004222  </td><td>2.049246e-02</td><td> 1          </td><td>  47        </td><td>24722       </td><td>116         </td><td>4.696189e-02</td></tr>\n",
       "\t<tr><td>GO:0031012  </td><td>2.499740e-03</td><td> 1          </td><td>  16        </td><td>24753       </td><td>116         </td><td>1.057582e-02</td></tr>\n",
       "\t<tr><td>GO:0000145  </td><td>4.308215e-03</td><td> 1          </td><td>  21        </td><td>24748       </td><td>116         </td><td>1.418667e-02</td></tr>\n",
       "\t<tr><td>GO:0006887  </td><td>3.529612e-03</td><td> 1          </td><td>  19        </td><td>24750       </td><td>116         </td><td>1.386633e-02</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllll}\n",
       " term & pval & genes\\_to\\_term\\_in\\_set & genes\\_to\\_term\\_in\\_genome & genes\\_not\\_term\\_in\\_genome & size\\_of\\_set & p.adj\\\\\n",
       "\\hline\n",
       "\t GO:0004568   & 5.158789e-03 &  1           &   23         & 24746        & 116          & 1.418667e-02\\\\\n",
       "\t GO:0006032   & 5.158789e-03 &  1           &   23         & 24746        & 116          & 1.418667e-02\\\\\n",
       "\t GO:0016998   & 5.158789e-03 &  1           &   23         & 24746        & 116          & 1.418667e-02\\\\\n",
       "\t GO:0006334   & 1.907271e-03 &  1           &   14         & 24755        & 116          & 9.536357e-03\\\\\n",
       "\t GO:0003677   & 2.781172e-06 & 17           & 1080         & 23689        & 116          & 3.824112e-05\\\\\n",
       "\t GO:0003700   & 8.756623e-13 & 16           &  355         & 24414        & 116          & 4.816142e-11\\\\\n",
       "\t GO:0006355   & 1.055418e-09 & 17           &  638         & 24131        & 116          & 2.902399e-08\\\\\n",
       "\t GO:0006457   & 3.855627e-03 &  2           &   67         & 24702        & 116          & 1.413730e-02\\\\\n",
       "\t GO:0051082   & 2.199133e-03 &  2           &   55         & 24714        & 116          & 1.007936e-02\\\\\n",
       "\t GO:0005509   & 2.167969e-06 &  7           &  180         & 24589        & 116          & 3.824112e-05\\\\\n",
       "\t GO:0005388   & 7.070310e-03 &  1           &   27         & 24742        & 116          & 1.690726e-02\\\\\n",
       "\t GO:0070588   & 6.566561e-03 &  1           &   26         & 24743        & 116          & 1.641640e-02\\\\\n",
       "\t GO:0005618   & 2.676722e-05 &  3           &   37         & 24732        & 116          & 2.453662e-04\\\\\n",
       "\t GO:0006073   & 2.676722e-05 &  3           &   37         & 24732        & 116          & 2.453662e-04\\\\\n",
       "\t GO:0016762   & 3.309550e-05 &  3           &   39         & 24730        & 116          & 2.600361e-04\\\\\n",
       "\t GO:0048046   & 1.571080e-03 &  3           &  106         & 24663        & 116          & 8.640942e-03\\\\\n",
       "\t GO:0010411   & 5.387248e-04 &  2           &   34         & 24735        & 116          & 3.703733e-03\\\\\n",
       "\t GO:0042546   & 6.384324e-04 &  2           &   36         & 24733        & 116          & 3.901532e-03\\\\\n",
       "\t GO:0016779   & 4.724593e-03 &  1           &   22         & 24747        & 116          & 1.418667e-02\\\\\n",
       "\t GO:0008237   & 5.610630e-03 &  1           &   24         & 24745        & 116          & 1.469451e-02\\\\\n",
       "\t GO:0004222   & 2.049246e-02 &  1           &   47         & 24722        & 116          & 4.696189e-02\\\\\n",
       "\t GO:0031012   & 2.499740e-03 &  1           &   16         & 24753        & 116          & 1.057582e-02\\\\\n",
       "\t GO:0000145   & 4.308215e-03 &  1           &   21         & 24748        & 116          & 1.418667e-02\\\\\n",
       "\t GO:0006887   & 3.529612e-03 &  1           &   19         & 24750        & 116          & 1.386633e-02\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| term | pval | genes_to_term_in_set | genes_to_term_in_genome | genes_not_term_in_genome | size_of_set | p.adj |\n",
       "|---|---|---|---|---|---|---|\n",
       "| GO:0004568   | 5.158789e-03 |  1           |   23         | 24746        | 116          | 1.418667e-02 |\n",
       "| GO:0006032   | 5.158789e-03 |  1           |   23         | 24746        | 116          | 1.418667e-02 |\n",
       "| GO:0016998   | 5.158789e-03 |  1           |   23         | 24746        | 116          | 1.418667e-02 |\n",
       "| GO:0006334   | 1.907271e-03 |  1           |   14         | 24755        | 116          | 9.536357e-03 |\n",
       "| GO:0003677   | 2.781172e-06 | 17           | 1080         | 23689        | 116          | 3.824112e-05 |\n",
       "| GO:0003700   | 8.756623e-13 | 16           |  355         | 24414        | 116          | 4.816142e-11 |\n",
       "| GO:0006355   | 1.055418e-09 | 17           |  638         | 24131        | 116          | 2.902399e-08 |\n",
       "| GO:0006457   | 3.855627e-03 |  2           |   67         | 24702        | 116          | 1.413730e-02 |\n",
       "| GO:0051082   | 2.199133e-03 |  2           |   55         | 24714        | 116          | 1.007936e-02 |\n",
       "| GO:0005509   | 2.167969e-06 |  7           |  180         | 24589        | 116          | 3.824112e-05 |\n",
       "| GO:0005388   | 7.070310e-03 |  1           |   27         | 24742        | 116          | 1.690726e-02 |\n",
       "| GO:0070588   | 6.566561e-03 |  1           |   26         | 24743        | 116          | 1.641640e-02 |\n",
       "| GO:0005618   | 2.676722e-05 |  3           |   37         | 24732        | 116          | 2.453662e-04 |\n",
       "| GO:0006073   | 2.676722e-05 |  3           |   37         | 24732        | 116          | 2.453662e-04 |\n",
       "| GO:0016762   | 3.309550e-05 |  3           |   39         | 24730        | 116          | 2.600361e-04 |\n",
       "| GO:0048046   | 1.571080e-03 |  3           |  106         | 24663        | 116          | 8.640942e-03 |\n",
       "| GO:0010411   | 5.387248e-04 |  2           |   34         | 24735        | 116          | 3.703733e-03 |\n",
       "| GO:0042546   | 6.384324e-04 |  2           |   36         | 24733        | 116          | 3.901532e-03 |\n",
       "| GO:0016779   | 4.724593e-03 |  1           |   22         | 24747        | 116          | 1.418667e-02 |\n",
       "| GO:0008237   | 5.610630e-03 |  1           |   24         | 24745        | 116          | 1.469451e-02 |\n",
       "| GO:0004222   | 2.049246e-02 |  1           |   47         | 24722        | 116          | 4.696189e-02 |\n",
       "| GO:0031012   | 2.499740e-03 |  1           |   16         | 24753        | 116          | 1.057582e-02 |\n",
       "| GO:0000145   | 4.308215e-03 |  1           |   21         | 24748        | 116          | 1.418667e-02 |\n",
       "| GO:0006887   | 3.529612e-03 |  1           |   19         | 24750        | 116          | 1.386633e-02 |\n",
       "\n"
      ],
      "text/plain": [
       "   term       pval         genes_to_term_in_set genes_to_term_in_genome\n",
       "1  GO:0004568 5.158789e-03  1                     23                   \n",
       "2  GO:0006032 5.158789e-03  1                     23                   \n",
       "3  GO:0016998 5.158789e-03  1                     23                   \n",
       "4  GO:0006334 1.907271e-03  1                     14                   \n",
       "5  GO:0003677 2.781172e-06 17                   1080                   \n",
       "6  GO:0003700 8.756623e-13 16                    355                   \n",
       "7  GO:0006355 1.055418e-09 17                    638                   \n",
       "8  GO:0006457 3.855627e-03  2                     67                   \n",
       "9  GO:0051082 2.199133e-03  2                     55                   \n",
       "10 GO:0005509 2.167969e-06  7                    180                   \n",
       "11 GO:0005388 7.070310e-03  1                     27                   \n",
       "12 GO:0070588 6.566561e-03  1                     26                   \n",
       "13 GO:0005618 2.676722e-05  3                     37                   \n",
       "14 GO:0006073 2.676722e-05  3                     37                   \n",
       "15 GO:0016762 3.309550e-05  3                     39                   \n",
       "16 GO:0048046 1.571080e-03  3                    106                   \n",
       "17 GO:0010411 5.387248e-04  2                     34                   \n",
       "18 GO:0042546 6.384324e-04  2                     36                   \n",
       "19 GO:0016779 4.724593e-03  1                     22                   \n",
       "20 GO:0008237 5.610630e-03  1                     24                   \n",
       "21 GO:0004222 2.049246e-02  1                     47                   \n",
       "22 GO:0031012 2.499740e-03  1                     16                   \n",
       "23 GO:0000145 4.308215e-03  1                     21                   \n",
       "24 GO:0006887 3.529612e-03  1                     19                   \n",
       "   genes_not_term_in_genome size_of_set p.adj       \n",
       "1  24746                    116         1.418667e-02\n",
       "2  24746                    116         1.418667e-02\n",
       "3  24746                    116         1.418667e-02\n",
       "4  24755                    116         9.536357e-03\n",
       "5  23689                    116         3.824112e-05\n",
       "6  24414                    116         4.816142e-11\n",
       "7  24131                    116         2.902399e-08\n",
       "8  24702                    116         1.413730e-02\n",
       "9  24714                    116         1.007936e-02\n",
       "10 24589                    116         3.824112e-05\n",
       "11 24742                    116         1.690726e-02\n",
       "12 24743                    116         1.641640e-02\n",
       "13 24732                    116         2.453662e-04\n",
       "14 24732                    116         2.453662e-04\n",
       "15 24730                    116         2.600361e-04\n",
       "16 24663                    116         8.640942e-03\n",
       "17 24735                    116         3.703733e-03\n",
       "18 24733                    116         3.901532e-03\n",
       "19 24747                    116         1.418667e-02\n",
       "20 24745                    116         1.469451e-02\n",
       "21 24722                    116         4.696189e-02\n",
       "22 24753                    116         1.057582e-02\n",
       "23 24748                    116         1.418667e-02\n",
       "24 24750                    116         1.386633e-02"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "go_results_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: Problem with `filter()` input `..1`.\ni Input `..1` is `Category == \"BiologicalProcess\"`.\nx object 'Category' not found\n",
     "output_type": "error",
     "traceback": [
      "Error: Problem with `filter()` input `..1`.\ni Input `..1` is `Category == \"BiologicalProcess\"`.\nx object 'Category' not found\nTraceback:\n",
      "1. go_results_1 %>% filter(Category == \"BiologicalProcess\")",
      "2. filter(., Category == \"BiologicalProcess\")",
      "3. filter.data.frame(., Category == \"BiologicalProcess\")",
      "4. filter_rows(.data, ..., caller_env = caller_env())",
      "5. withCallingHandlers({\n .     dots <- new_quosures(flatten(imap(dots, function(dot, index) {\n .         env_filter$current_expression <- index\n .         expand_if_across(dot)\n .     })))\n .     mask$eval_all_filter(dots, env_filter)\n . }, error = function(e) {\n .     local_call_step(dots = dots, .index = env_filter$current_expression, \n .         .fn = \"filter\")\n .     abort(c(cnd_bullet_header(), i = cnd_bullet_input_info(), \n .         x = conditionMessage(e), i = cnd_bullet_cur_group_label()), \n .         class = \"dplyr_error\")\n . })",
      "6. mask$eval_all_filter(dots, env_filter)",
      "7. .handleSimpleError(function (e) \n . {\n .     local_call_step(dots = dots, .index = env_filter$current_expression, \n .         .fn = \"filter\")\n .     abort(c(cnd_bullet_header(), i = cnd_bullet_input_info(), \n .         x = conditionMessage(e), i = cnd_bullet_cur_group_label()), \n .         class = \"dplyr_error\")\n . }, \"object 'Category' not found\", base::quote(mask$eval_all_filter(dots, \n .     env_filter)))",
      "8. h(simpleError(msg, call))",
      "9. abort(c(cnd_bullet_header(), i = cnd_bullet_input_info(), x = conditionMessage(e), \n .     i = cnd_bullet_cur_group_label()), class = \"dplyr_error\")",
      "10. signal_abort(cnd)"
     ]
    }
   ],
   "source": [
    "# example of use, see also go_enrichment_test\n",
    "go1 <- go_results_1 %>% filter(Category == \"BiologicalProcess\")\n",
    "# go2 <- go_results_2 %>% filter(Category == \"BiologicalProcess\")\n",
    "# compare_two(as.character(go_results_1[['term']]), as.character(go_results_2[['term']]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
